lib/Conversion/ImportVerilog/Expressions.cpp

```diff
@@ -361,12 +361,56 @@ struct ExprVisitor {
     for (auto *operand : expr.operands()) {
       auto value = context.convertExpression(*operand);
       if (!value)
-        return {};
+        continue;
+      value = convertToSimpleBitVector(value);
       operands.push_back(value);
     }
     return builder.create<moore::ConcatOp>(loc, operands);
   }
 
+  // Handle replications.
+  Value visit(const slang::ast::ReplicationExpression &expr) {
+    auto type = context.convertType(*expr.type);
+    if (isa<moore::VoidType>(type))
+      return {};
+
+    auto value = context.convertExpression(expr.concat());
+    if (!value)
+      return {};
+    return builder.create<moore::ReplicateOp>(loc, type, value);
+  }
+
+  // Handle single bit selections.
+  Value visit(const slang::ast::ElementSelectExpression &expr) {
+    auto type = context.convertType(*expr.type);
+    auto value = context.convertExpression(expr.value());
+    auto lowBit = context.convertExpression(expr.selector());
+
+    if (!value || !lowBit)
+      return {};
+    return builder.create<moore::ExtractOp>(loc, type, value, lowBit);
+  }
+
+  // Handle range bits selections.
+  Value visit(const slang::ast::RangeSelectExpression &expr) {
+    auto type = context.convertType(*expr.type);
+    auto value = context.convertExpression(expr.value());
+    Value lowBit;
+    if (expr.getSelectionKind() == slang::ast::RangeSelectionKind::Simple) {
+      if (expr.left().constant && expr.right().constant) {
+        auto lhs = expr.left().constant->integer().as<uint64_t>().value();
+        auto rhs = expr.right().constant->integer().as<uint64_t>().value();
+        lowBit = lhs < rhs ? context.convertExpression(expr.left())
+                           : context.convertExpression(expr.right());
+      }
+    } else
+      lowBit = context.convertExpression(expr.left());
```

Comment:
What happens with the right-hand side expression? If I write `x[a:b]`, this just stores `x` and `a` into the op, but `b` gets lost. Shouldn't that go somewhere in the op? I think it would be good to just reject all the cases here that we can't really fully handle. It's totally fine to say that we only support constant left/right expressions for now. Or that we only support `[a+:k]` or `[a-:k]` if it has a constant k expression. We then probably want different ops for the different cases (extract with two constants, extract with constant width but dynamic offset, extract where both sides are dynamic -- probably very last).

Reply:
In Slang AST, the RHS like `k` just represents the width, we don't need to capture it(must be a positive constant, and x/z/non-constant will trigger errors).
In your case x[a:b](belong to `RangeSelectionKind::Simple`), firstly, `a` and `b` must be constant, then they are meaningful to capture to distinguish big and little endian like `[0:3]` and `[3:0]`. When we select a range of bits like `[1:2]` from `[0:3]` and `[3:2]` from `[3:0]`, the `lowBit` is`1` and `2` separately.
`else` statement serves at `RangeSelectionKind::IndexedUp` and `RangeSelectionKind::IndexedDown`.

Reply:
I'll modify the git message for `moore.extract`, it's not accurate right now.